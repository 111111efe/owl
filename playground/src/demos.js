export const HELLO_WORLD = `class HelloWorld extends owl.core.Component {
  inlineTemplate = \`<div>Hello <t t-esc="props.name"/></div>\`;
}

const env = {
    qweb: new owl.core.QWeb()
};

const hello = new HelloWorld(env, { name: "World" });
hello.mount(document.body);
`;
