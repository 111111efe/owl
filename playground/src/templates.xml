<templates>
  <div t-name="tabbed-editor" class="tabbed-editor" t-att-style="props.style">
    <div class="tabBar" t-att-class="{resizeable: props.resizeable}" t-on-mousedown="onMouseDown">
      <a t-if="tabs.js" class="tab" t-att-class="{active: state.currentTab==='js'}" t-on-click="setTab('js')">JS</a>
      <a t-if="tabs.xml" class="tab" t-att-class="{active: state.currentTab==='xml'}" t-on-click="setTab('xml')">XML</a>
      <a t-if="tabs.css" class="tab" t-att-class="{active: state.currentTab==='css'}" t-on-click="setTab('css')">CSS</a>
    </div>
    <div class="code-editor" t-ref="'editor'"></div>
  </div>

  <div t-name="playground" class="playground">
      <div class="left-bar" t-att-style="leftPaneStyle" t-att-class="{split: state.splitLayout}">
        <div class="menubar">
          <a class="btn run-code" t-on-click="runCode">â–¶ Run</a>
          <select t-on-change="setSample">
            <option t-foreach="SAMPLES" t-as="sample" t-key="sample_index">
              <t t-esc="sample.description"/>
            </option>
          </select>
          <div class="layout-selector" t-att-class="{active:state.splitLayout}" t-on-click="toggleLayout">â—«</div>
        </div>
        <t t-if="!state.splitLayout">
          <t t-widget="TabbedEditor" t-props="{js:state.js, css:state.css, xml: state.xml, display: 'js|xml|css'}" t-on-updateCode="updateCode"/>
        </t>
        <t t-else="1">
          <t t-widget="TabbedEditor" t-props="{js:state.js, css:state.css, xml: state.xml, display: 'js', style:topEditorStyle}" t-on-updateCode="updateCode"/>
          <div class="separator horizontal"/>
          <t t-widget="TabbedEditor" t-keepalive="1" t-props="{js:state.js, css:state.css, xml: state.xml, display: 'xml|css', resizeable: true}" t-on-updateCode="updateCode" t-on-updatePanelHeight="updatePanelHeight"/>
        </t>
      </div>
      <div class="separator vertical" t-on-mousedown="onMouseDown"/>
      <div class="right-pane"  t-att-style="rightPaneStyle">
        <div class="welcome" t-if="state.displayWelcome">
          <div>ðŸ¦‰ Odoo Web Library ðŸ¦‰</div>
          <div>v<t t-esc="version"/></div>
          <div class="url"><a href="https://github.com/odoo/owl">https://github.com/odoo/owl</a></div>
          <div class="note">Note: these examples require a recent browser to work without a transpilation step. </div>
        </div>
        <div t-if="state.error" class="error">
          <t t-esc="state.error"/>
        </div>
        <div class="content" t-ref="'content'"/>
      </div>
  </div>
</templates>