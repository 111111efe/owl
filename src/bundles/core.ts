import QWeb from "../core/qweb_vdom";
import * as utils from "../core/utils";
import Widget from "../core/widget";


(<any>window).core = {
    QWeb,
    utils,
    Widget
}