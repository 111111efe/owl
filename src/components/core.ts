import { qweb } from "../qweb/qweb";
import { QWebTemplate } from "../qweb/types";
import { VRootNode } from "../vdom/types";
import { buildTree, patch } from "../vdom/vdom";
import { Component } from "./component";
import { Fiber } from "./fiber";
import { scheduler } from "./scheduler";

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

export interface FComponent<T = any, Env = any> {
  template: string;
  components?: { [key: string]: FComponent<any> | CComponent };
  name?: string;
  setup?: (props: any, env: Env) => T | Promise<T> | void;
}

export type CComponent = typeof Component;

export type CInstance<Props = any, Env = any> = Component<Props, Env>;

export type FInstance<T = {}, Env = any> = T & {
  __owl__: OwlElement<T>;
  props: any;
  env: Env;
};

export interface OwlElement<T = any> {
  fiber: Fiber | null;
  name: string;
  components: { [key: string]: FComponent<T> | CComponent };
  isMounted: boolean;
  instance: FInstance<T> | CInstance | null;
  vnode: VRootNode | null;
  qweb: QWebTemplate;
  isReady: Promise<void> | null;
  children: { [key: string]: OwlElement };
}

// -----------------------------------------------------------------------------
// core object
// -----------------------------------------------------------------------------
export const core = {
  current: null as null | OwlElement,
};

// -----------------------------------------------------------------------------
// Mount
// -----------------------------------------------------------------------------
export type MountTarget = HTMLElement | DocumentFragment;

export interface MountConfig<Env = any> {
  props?: Object;
  env?: Env;
}

export function mount<C extends CComponent>(
  Comp: C,
  target: MountTarget,
  config?: MountConfig
): Promise<InstanceType<C>>;
export function mount(
  comp: CInstance,
  target: MountTarget,
  config?: MountConfig
): Promise<CInstance>;
export function mount<T, Env>(
  Comp: FComponent<T>,
  target: MountTarget,
  config?: MountConfig<Env>
): Promise<FInstance<T extends {} ? T : {}, Env>>;
export function mount<T, Env>(
  comp: FInstance,
  target: MountTarget,
  config?: MountConfig<Env>
): Promise<FInstance<T extends {} ? T : {}, Env>>;
export async function mount(
  info: any,
  target: MountTarget,
  config: MountConfig = {}
): Promise<CInstance | FInstance> {
  if (!(target instanceof HTMLElement || target instanceof DocumentFragment)) {
    const name =
      info instanceof Component ? info.constructor.name : (info as any).name || "No Name";
    let message = `Component '${name}' cannot be mounted: the target is not a valid DOM node.`;
    message += `\nMaybe the DOM is not ready yet? (in that case, you can use owl.utils.whenReady)`;
    throw new Error(message);
  }
  if (info instanceof Component) {
    // this is a CInstance
    const elem = info.__owl__!;
    return scheduler.addFiber(elem.fiber!).then(() => {
      target.appendChild(info.el!);
      return info;
    });
  }
  let element: OwlElement;
  let props = config.props || {};
  let env = config.env || {};
  if (info.prototype instanceof Component || info === Component) {
    // this is a CComponent (Class Component)
    element = createCComponent(info, props, env);
  } else if ("__owl__" in info) {
    // this is a CInstance or a FInstance
    element = info.__owl__;
  } else {
    // this is a FComponent
    element = createFComponent(info, props, env);
  }
  return scheduler.addFiber(element.fiber!).then(() => {
    const fragment = document.createDocumentFragment();
    buildTree(element.vnode!, (fragment as any) as HTMLElement);
    target.appendChild(fragment);
    if (document.body.contains(target)) {
      mountTree(element);
    }
    return element.instance!;
  });
}

function mountTree(element: OwlElement) {
  for (let k in element.children) {
    mountTree(element.children[k]);
  }
  element.isMounted = true;
  if (element.instance.mounted) {
    element.instance.mounted();
  }
}

// -----------------------------------------------------------------------------
// Component creation
// -----------------------------------------------------------------------------

export function createFComponent(fn: FComponent, props: any, env: any): OwlElement {
  let template: string = fn.template;
  const name = fn.name || "Anonymous Function Component";
  if (!template) {
    throw new Error(`Component "${name}" does not have a template defined!`);
  }
  const qtemplate = qweb.getTemplate(template);
  const elem: OwlElement = {
    fiber: null,
    name,
    instance: null,
    isMounted: false,
    vnode: null,
    components: fn.components || {},
    qweb: qtemplate,
    isReady: null,
    children: {},
  };
  core.current = elem;

  if (fn.setup) {
    const result = fn.setup(props, env);
    const prom = result instanceof Promise ? result : Promise.resolve(result);
    elem.isReady = prom.then((result) => {
      let instance = result ? Object.create(result) : {};
      instance.props = props;
      instance.__owl__ = elem;
      instance.env = env;
      elem.instance = instance;
    });
  } else {
    const instance: FInstance = { props, env, __owl__: elem };
    elem.instance = instance;
    elem.isReady = Promise.resolve();
  }

  startComponent(elem);
  return elem;
}

export function createCComponent<C extends typeof Component>(
  C: C,
  props: any,
  env: any
): OwlElement {
  let template: string = C.template;
  if (!template) {
    throw new Error(`Component "${C.name}" does not have a template defined!`);
  }
  const qtemplate = qweb.getTemplate(template);
  const elem: OwlElement = {
    fiber: null,
    name: C.name,
    instance: null,
    isMounted: false,
    vnode: null,
    components: C.components,
    qweb: qtemplate,
    isReady: null,
    children: {},
  };
  core.current = elem;
  const component = new C(props, env) as InstanceType<C>;
  component.__owl__ = elem;
  elem.instance = component;
  elem.isReady = component.willStart();
  startComponent(elem);
  return elem;
}

// -----------------------------------------------------------------------------
// starting Components
// -----------------------------------------------------------------------------

async function startComponent(elem: OwlElement) {
  const vnode = elem.qweb.createRoot();
  vnode.hooks.create = (el) => (elem.instance.el = el);
  const fiber = new Fiber(elem, null, vnode);
  fiber.root.counter++;
  elem.fiber = fiber;
  elem.vnode = vnode;
  await elem.isReady!;
  elem.qweb.render(vnode, elem.instance, elem);
  fiber.root.counter--;
}

export function renderToFiber(elem: OwlElement, parentFiber: Fiber | null): Fiber {
  const newRoot = elem.qweb.createRoot();
  const fiber = new Fiber(elem, parentFiber, newRoot);
  elem.fiber = fiber;
  fiber.root.counter++;
  new Promise((resolve) => {
    elem.qweb.render(newRoot, elem.instance!, elem);
    fiber.root.counter--;
    resolve();
  });
  return fiber;
}
// -----------------------------------------------------------------------------
// render
// -----------------------------------------------------------------------------

export function render(elem: OwlElement): Promise<void> {
  let fiber = renderToFiber(elem, null);
  return scheduler.addFiber(fiber).then(() => {
    patch(elem.vnode!, fiber.vnode!);
    while (fiber.next) {
      fiber = fiber.next;
      patch(fiber.elem.vnode!, fiber.vnode!);
    }
  });
}
