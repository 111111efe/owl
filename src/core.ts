import { BDom, Block, Blocks } from "./bdom";
import { TemplateSet } from "./qweb_compiler";
import { observe } from "./reactivity";

// -----------------------------------------------------------------------------
//  Global templates
// -----------------------------------------------------------------------------

let nextId = 1;
export const globalTemplates = new TemplateSet();

export function xml(strings: TemplateStringsArray, ...args: any[]) {
  const name = `__template__${nextId++}`;
  const value = String.raw(strings, ...args);
  globalTemplates.add(name, value);
  return name;
}

// -----------------------------------------------------------------------------
//  Component
// -----------------------------------------------------------------------------

interface ComponentData {
  bdom: null | BDom;
  render: () => BDom;
  fiber: Fiber | null;
  willStartCB: Function;
  mountedCB: Function;
  isMounted: boolean;
  children: { [key: string]: Component };
  slots?: any;
}

export class Component {
  static template: string;
  __owl__: ComponentData = currentData;
  props: any;
  env: any = currentEnv;

  constructor(props: any) {
    current = this;
    const __owl__ = currentData;
    __owl__.willStartCB = this.willStart.bind(this);
    __owl__.mountedCB = this.mounted.bind(this);
    this.props = props;
  }

  setup() {}
  async willStart(): Promise<void> {}
  async willUpdateProps(props: any): Promise<void> {}
  mounted() {}

  get el(): ChildNode | null {
    const bdom = this.__owl__.bdom;
    return bdom ? bdom.el : null;
  }

  async render(): Promise<void> {
    const fiber = new RootFiber(this.__owl__!);
    scheduler.addFiber(fiber);
    internalRender(this, fiber);
    await fiber.promise;
  }
}

// -----------------------------------------------------------------------------
//  Component Block
// -----------------------------------------------------------------------------

class BComponent extends Block {
  component: Component;
  handlers?: any[];

  constructor(name: string, props: any, key: string, ctx: any) {
    super();
    const parentData: ComponentData = ctx.__owl__;
    let component = parentData.children[key];
    if (component) {
      // update
      const fiber = new ChildFiber(component.__owl__, parentData.fiber!);
      const parentFiber = parentData.fiber!;
      parentFiber.child = fiber; // wrong!
      updateAndRender(component, fiber, props);
    } else {
      // new component
      const components = ctx.constructor.components || ctx.components;
      const C = components[name];
      component = prepare(C, props);
      parentData.children[key] = component;
      const fiber = new ChildFiber(component.__owl__, parentData.fiber!);
      const parentFiber = parentData.fiber!;
      parentFiber.child = fiber; // wrong!
      internalRender(component, fiber);
    }
    this.component = component;
  }

  firstChildNode(): ChildNode | null {
    const bdom = this.component.__owl__.bdom;
    return bdom ? bdom.firstChildNode() : null;
  }

  mountBefore(anchor: ChildNode) {
    this.component.__owl__!.bdom = this.component.__owl__!.fiber!.bdom;
    this.component.__owl__!.bdom!.mountBefore(anchor);
  }
  patch() {
    this.component.__owl__!.bdom!.patch(this.component.__owl__!.fiber!.bdom);
  }
}

export class BComponentH extends BComponent {
  handlers: any[];
  constructor(handlers: number, name: string, props: any, key: string, ctx: any) {
    super(name, props, key, ctx);
    this.handlers = new Array(handlers);
  }

  mountBefore(anchor: ChildNode) {
    super.mountBefore(anchor);
    this.setupHandlers();
  }

  setupHandlers() {
    for (let i = 0; i < this.handlers.length; i++) {
      const handler = this.handlers[i];
      const eventType = handler[0];
      const el = this.component.el!;
      el.addEventListener(eventType, () => {
        const info = this.handlers![i];
        const [, callback, ctx] = info;
        if (ctx.__owl__ && !ctx.__owl__.isMounted) {
          return;
        }
        callback();
      });
    }
  }
}

async function updateAndRender(component: Component, fiber: ChildFiber, props: any) {
  const componentData = component.__owl__;
  componentData.fiber = fiber;
  await component.willUpdateProps(props);
  component.props = props;
  fiber.bdom = componentData.render();
  fiber.root.counter--;
}

Blocks.BComponent = BComponent;
Blocks.BComponentH = BComponentH;

// -----------------------------------------------------------------------------
//  Scheduler
// -----------------------------------------------------------------------------

export class Scheduler {
  tasks: RootFiber[] = [];
  isRunning: boolean = false;
  requestAnimationFrame: Window["requestAnimationFrame"];

  constructor(requestAnimationFrame: Window["requestAnimationFrame"]) {
    this.requestAnimationFrame = requestAnimationFrame;
  }

  start() {
    this.isRunning = true;
    this.scheduleTasks();
  }

  stop() {
    this.isRunning = false;
  }

  addFiber(fiber: RootFiber) {
    this.tasks.push(fiber);
    if (!this.isRunning) {
      this.start();
    }
  }

  /**
   * Process all current tasks. This only applies to the fibers that are ready.
   * Other tasks are left unchanged.
   */
  flush() {
    let tasks = this.tasks;
    this.tasks = [];
    tasks = tasks.filter((fiber) => {
      if (fiber.counter === 0) {
        if (!fiber.error) {
          fiber.complete();
        }
        fiber.resolve();
        return false;
      }
      return true;
    });
    this.tasks = tasks.concat(this.tasks);
    if (this.tasks.length === 0) {
      this.stop();
    }
  }

  scheduleTasks() {
    this.requestAnimationFrame(() => {
      this.flush();
      if (this.isRunning) {
        this.scheduleTasks();
      }
    });
  }
}

const scheduler = new Scheduler(window.requestAnimationFrame.bind(window));

// -----------------------------------------------------------------------------
//  Internal rendering stuff
// -----------------------------------------------------------------------------

class BaseFiber {
  bdom: BDom | null = null;
  error?: Error;
  __owl__: ComponentData;

  child: Fiber | null = null;
  sibling: Fiber | null = null;

  constructor(__owl__: ComponentData) {
    this.__owl__ = __owl__;
  }

  mountComponents() {
    if (this.child) {
      this.child.mountComponents();
    }
    if (this.sibling) {
      this.sibling.mountComponents();
    }
    this.__owl__.mountedCB();
    this.__owl__.isMounted = true;
  }
}

type Fiber = ChildFiber | RootFiber;

class ChildFiber extends BaseFiber {
  bdom: BDom | null = null;
  error?: Error;
  root: RootFiber;
  parent: BaseFiber;

  constructor(__owl__: ComponentData, parent: Fiber) {
    super(__owl__);
    this.parent = parent;
    const root = parent.root;
    root.counter++;
    root.childNumber++;
    this.root = root;
  }
}

class RootFiber extends BaseFiber {
  counter: number = 1;
  childNumber: number = 1;
  root: RootFiber = this;

  resolve!: () => void;
  reject!: (error: Error) => void;
  promise = new Promise((resolve, reject) => {
    this.resolve = resolve;
    this.reject = reject;
  });

  complete() {
    this.__owl__!.bdom!.patch(this.bdom);
  }
}

class MountingFiber extends RootFiber {
  target: HTMLElement | DocumentFragment;

  constructor(__owl__: ComponentData, target: HTMLElement | DocumentFragment) {
    super(__owl__);
    this.target = target;
  }
  complete() {
    const __owl__ = this.__owl__!;
    __owl__.bdom! = __owl__.fiber!.bdom!;
    __owl__.bdom!.mount(this.target);
    if (document.body.contains(this.target)) {
      this.mountComponents();
    }
  }
}

type Env = any;

interface MountParameters {
  env?: Env;
  target: HTMLElement | DocumentFragment;
  props?: any;
}

interface Type<T> extends Function {
  new (...args: any[]): T;
}

let current: Component | null = null;
let currentData: ComponentData;
let currentEnv: any;

export function mount<T extends Type<Component>>(
  C: T | Component,
  params: MountParameters
): Promise<InstanceType<T>>;
export async function mount(C: any, params: MountParameters) {
  if (!(params.target instanceof HTMLElement || params.target instanceof DocumentFragment)) {
    throw new Error("Cannot mount component: the target is not a valid DOM element");
  }
  if (C instanceof Component) {
    // we move component elsewhere
    C.__owl__.bdom!.move(params.target);
    return;
  }
  const { target, props, env } = params;
  currentEnv = env || {};
  const component = prepare(C, props || {});
  const fiber = new MountingFiber(component.__owl__!, target);
  scheduler.addFiber(fiber);
  internalRender(component, fiber);
  await fiber.promise;
  return component;
}

function prepare(C: any, props: any): Component {
  let component: Component;
  let template: string = (C as any).template;
  if (!template) {
    throw new Error(`Could not find template for component "${C.name}"`);
  }
  const __owl__: ComponentData = {
    render: null as any,
    bdom: null,
    fiber: null,
    willStartCB: null as any,
    mountedCB: null as any,
    isMounted: false,
    children: {},
  };
  currentData = __owl__;

  component = new C(props);
  component.setup();
  __owl__.render = globalTemplates.getFunction(template).bind(null, component);
  return component;
}

async function internalRender(c: Component, fiber: Fiber) {
  const __owl__ = c.__owl__!;
  __owl__.fiber = fiber;
  await __owl__.willStartCB();
  fiber.bdom = __owl__.render();
  fiber.root.counter--;
}

// -----------------------------------------------------------------------------
//  hooks
// -----------------------------------------------------------------------------

export function useState<T>(state: T): T {
  const component: Component = current!;
  return observe(state, () => component.render());
}

export function onWillStart(cb: any) {
  const component: Component = current!;
  const prev = currentData!.willStartCB;
  currentData!.willStartCB = () => {
    return Promise.all([prev.call(component), cb.call(component)]);
  };
}

export function onMounted(cb: any) {
  const component: Component = current!;
  const prev = currentData!.mountedCB;
  currentData!.mountedCB = () => {
    prev();
    cb.call(component);
  };
}

// -----------------------------------------------------------------------------
// useRef
// -----------------------------------------------------------------------------

/**
 * The purpose of this hook is to allow components to get a reference to a sub
 * html node or component.
 */
interface Ref<C extends Component = Component> {
  el: HTMLElement | null;
  comp: C | null;
}

export function useRef<C extends Component = Component>(name: string): Ref<C> {
  const __owl__ = currentData!;
  return {
    get el(): HTMLElement | null {
      const val = __owl__.bdom && __owl__.bdom.refs && __owl__.bdom.refs[name];
      return val!;
      // if (val instanceof HTMLElement) {
      //   return val;
      // } else if (val instanceof Component) {
      //   return val.el;
      // }
      // return null;
    },
    get comp(): C | null {
      return null;
      // const val = __owl__.refs && __owl__.refs[name];
      // return val instanceof Component ? (val as C) : null;
    },
  };
}
