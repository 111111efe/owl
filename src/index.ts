import { QWeb } from "./qweb";
import { EventBus } from "./event_bus";
import { Component, PureComponent } from "./component";
import { Store, connect } from "./store";
import * as utils from "./utils";
export const core = {
  QWeb,
  EventBus,
  Component,
  PureComponent,
  Store,
  connect,
  utils
};
