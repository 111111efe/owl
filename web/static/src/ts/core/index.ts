export { QWeb } from "./qweb_vdom";
export { EventBus } from "./event_bus";
export { Component } from "./component";
