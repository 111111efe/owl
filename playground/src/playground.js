import { HELLO_WORLD } from "./demos.js";

class Playground extends owl.core.Component {
  inlineTemplate = `
    <div class="playground">
        <div class="left-bar">
            <button t-on-click="runCode">â–¶ Run</button>
            <div class="code-editor" t-ref="editor"></div>
        </div>
        <div class="content">
          <iframe t-ref="iframe"/>
        </div>
    </div>`;

  mounted() {
    this.editor = ace.edit(this.refs.editor);
    this.editor.session.setOption("useWorker", false);
    this.editor.setValue(HELLO_WORLD, -1);
    this.editor.setTheme("ace/theme/monokai");
    this.editor.session.setMode("ace/mode/javascript");
    const doc = this.refs.iframe.contentWindow.document;

    // add owl library
    const owlScript = doc.createElement("script");
    owlScript.src = "libs/owl.js";
    doc.body.appendChild(owlScript);
  }

  runCode() {
    const doc = this.refs.iframe.contentWindow.document;

    // // add owl library
    // const owlScript = doc.createElement("script");
    // owlScript.src = "libs/owl.js";
    // doc.body.appendChild(owlScript);

    const code = this.editor.getValue();
    const script = doc.createElement("script");
    script.type = "text/javascript";
    script.innerHTML = code;
    doc.body.appendChild(script);
    //  var thisDoc = this.contentWindow.document;
    //     if ( ! thisDoc.getElementById(scriptID)) {
    //         var scriptObj = thisDoc.createElement("script");
    //         scriptObj.type = "text/javascript";
    //         scriptObj.id = scriptId;
    //         scriptObj.innerHTML = script;
    //         thisDoc.body.appendChild(scriptObj);
    //     }
    // });
  }
}

document.addEventListener("DOMContentLoaded", async function() {
  const env = {
    qweb: new owl.core.QWeb()
  };

  const playground = new Playground(env);
  playground.mount(document.body);
});
