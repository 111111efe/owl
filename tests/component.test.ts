import { fromName, makeTestFixture, snapshotTemplateCode } from "./helpers";
import { mount, Component, xml } from "../src/core";

let fixture: HTMLElement;

beforeEach(() => {
  fixture = makeTestFixture();
});

describe("basics", () => {
  test("can mount a simple class component", async () => {
    class Test extends Component {
      static template = xml`<span>simple vnode</span>`;
    }

    const component = await mount(Test, { target: fixture });
    expect(fixture.innerHTML).toBe("<span>simple vnode</span>");
    expect(component.el).toEqual(fixture.querySelector("span"));
    snapshotTemplateCode(Test.template);
  });

  test("can mount a simple class component", async () => {
    const Test = {
      template: xml`<span>simple vnode</span>`,
    };

    const component = await mount(Test, { target: fixture });
    expect(fixture.innerHTML).toBe("<span>simple vnode</span>");
    expect(component.el).toEqual(fixture.querySelector("span"));
    snapshotTemplateCode(fromName(Test.template));
  });

  test("simple functional component with text node", async () => {
    const Test = {
      template: xml`look, just text!`,
    };

    await mount(Test, { target: fixture });
    expect(fixture.innerHTML).toBe("look, just text!");
    snapshotTemplateCode(fromName(Test.template));
  });

  test("class component with dynamic text", async () => {
    class Test extends Component {
      static template = xml`<span>My value: <t t-esc="value"/></span>`;

      value = 42;
    }

    await mount(Test, { target: fixture });
    expect(fixture.innerHTML).toBe("<span>My value: 42</span>");
    snapshotTemplateCode(fromName(Test.template));
  });

  test("Multi root component", async () => {
    class Test extends Component {
      static template = xml`<span>1</span>text<span>2</span>`;

      value = 42;
    }

    await mount(Test, { target: fixture });
    expect(fixture.innerHTML).toBe(`<span>1</span>text<span>2</span>`);
    snapshotTemplateCode(fromName(Test.template));
  });
});
