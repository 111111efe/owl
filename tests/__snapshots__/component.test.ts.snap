// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composition sub widgets with some state rendered in a loop 1`] = `
"function anonymous(context,extra
) {
    let owner = context;
    context = Object.create(context);
    var h = this.utils.h;
    var c1 = [], p1 = {key:1};
    var vn1 = h('div', p1, c1);
    var _2 = context['state'].numbers;
    if (!_2) { throw new Error('QWeb error: Invalid loop expression')}
    if (typeof _2 === 'number') { _2 = Array.from(Array(_2).keys())}
    var _3 = _2 instanceof Array ? _2 : Object.keys(_2);
    var _4 = _2 instanceof Array ? _2 : Object.values(_2);
    for (let i = 0; i < _3.length; i++) {
        context.number_first = i === 0;
        context.number_last = i === _3.length - 1;
        context.number_parity = i % 2 === 0 ? 'even' : 'odd';
        context.number_index = i;
        context.number = _3[i];
        context.number_value = _4[i];
        //WIDGET
        let key8 = context['number'];
        let _5_index = c1.length;
        c1.push(null);
        let def6;
        let w7 = key8 in context.__widget__.cmap ? context.__widget__.children[context.__widget__.cmap[key8]] : false;
        let props7 = {};
        let isNew7 = !w7;
        if (w7 && w7.__widget__.renderPromise) {
            if (w7.__widget__.isStarted) {
                def6 = w7.updateProps(props7, extra.forceUpdate);
            } else {
                isNew7 = true
                if (props7 === w7.__widget__.renderProps) {
                    def6 = w7.__widget__.renderPromise;
                } else {
                    w7.destroy();
                    w7 = false
                }
            }
        }
        if (!def6) {
            if (w7) {
                def6 = w7.updateProps(props7, extra.forceUpdate);
            } else {
                w7 = new context.widgets['ChildWidget'](owner, props7);
                context.__widget__.cmap[key8] = w7.__widget__.id;
                def6 = w7._prepare();
            }
        }
        if (isNew7) {
            def6 = def6.then(vnode=>{let pvnode=h(vnode.sel, {key: key8});c1[_5_index]=pvnode;pvnode.data.hook = {insert(vn){let nvn=w7._mount(vnode, vn.elm);pvnode.elm=nvn.elm},remove(){w7.destroy()},destroy(){w7.destroy()}}; w7.__widget__.pvnode = pvnode;});
        } else {
            def6 = def6.then(()=>{if (w7.__widget__.isDestroyed) {return};let vnode;if (!w7.__widget__.vnode){vnode=w7.__widget__.pvnode} else { vnode=h(w7.__widget__.vnode.sel, {key: key8});vnode.elm=w7.el;vnode.data.hook = {insert(a){a.elm.parentNode.replaceChild(w7.el,a.elm);a.elm=w7.el;w7.__mount();},remove(){w7.destroy()}, destroy() {w7.destroy()}}}c1[_5_index]=vnode;});
        }
        extra.promises.push(def6);
    }
    return vn1;
}"
`;

exports[`random stuff/miscellaneous snapshotting compiled code 1`] = `
"function anonymous(context,extra
) {
    let owner = context;
    var h = this.utils.h;
    var c1 = [], p1 = {key:1};
    var vn1 = h('div', p1, c1);
    //WIDGET
    let key5 = \\"somestring\\";
    let _2_index = c1.length;
    c1.push(null);
    let def3;
    let w4 = key5 in context.__widget__.cmap ? context.__widget__.children[context.__widget__.cmap[key5]] : false;
    let props4 = {flag: context['state'].flag};
    let isNew4 = !w4;
    if (w4 && w4.__widget__.renderPromise) {
        if (w4.__widget__.isStarted) {
            def3 = w4.updateProps(props4, extra.forceUpdate);
        } else {
            isNew4 = true
            if (props4 === w4.__widget__.renderProps) {
                def3 = w4.__widget__.renderPromise;
            } else {
                w4.destroy();
                w4 = false
            }
        }
    }
    if (!def3) {
        if (w4) {
            def3 = w4.updateProps(props4, extra.forceUpdate);
        } else {
            w4 = new context.widgets['child'](owner, props4);
            context.__widget__.cmap[key5] = w4.__widget__.id;
            def3 = w4._prepare();
        }
    }
    if (isNew4) {
        def3 = def3.then(vnode=>{let pvnode=h(vnode.sel, {key: key5});c1[_2_index]=pvnode;pvnode.data.hook = {insert(vn){let nvn=w4._mount(vnode, vn.elm);pvnode.elm=nvn.elm},remove(){w4.destroy()},destroy(){w4.destroy()}}; w4.__widget__.pvnode = pvnode;});
    } else {
        def3 = def3.then(()=>{if (w4.__widget__.isDestroyed) {return};let vnode;if (!w4.__widget__.vnode){vnode=w4.__widget__.pvnode} else { vnode=h(w4.__widget__.vnode.sel, {key: key5});vnode.elm=w4.el;vnode.data.hook = {insert(a){a.elm.parentNode.replaceChild(w4.el,a.elm);a.elm=w4.el;w4.__mount();},remove(){w4.destroy()}, destroy() {w4.destroy()}}}c1[_2_index]=vnode;});
    }
    extra.promises.push(def3);
    return vn1;
}"
`;

exports[`random stuff/miscellaneous t-props should not be undefined (snapshotting) 1`] = `
"function anonymous(context,extra
) {
    let owner = context;
    var h = this.utils.h;
    var c1 = [], p1 = {key:1};
    var vn1 = h('div', p1, c1);
    //WIDGET
    let _2_index = c1.length;
    c1.push(null);
    let def3;
    let w4 = 4 in context.__widget__.cmap ? context.__widget__.children[context.__widget__.cmap[4]] : false;
    let props4 = {};
    let isNew4 = !w4;
    if (w4 && w4.__widget__.renderPromise) {
        if (w4.__widget__.isStarted) {
            def3 = w4.updateProps(props4, extra.forceUpdate);
        } else {
            isNew4 = true
            if (props4 === w4.__widget__.renderProps) {
                def3 = w4.__widget__.renderPromise;
            } else {
                w4.destroy();
                w4 = false
            }
        }
    }
    if (!def3) {
        if (w4) {
            def3 = w4.updateProps(props4, extra.forceUpdate);
        } else {
            w4 = new context.widgets['child'](owner, props4);
            context.__widget__.cmap[4] = w4.__widget__.id;
            def3 = w4._prepare();
        }
    }
    if (isNew4) {
        def3 = def3.then(vnode=>{let pvnode=h(vnode.sel, {key: 4});c1[_2_index]=pvnode;pvnode.data.hook = {insert(vn){let nvn=w4._mount(vnode, vn.elm);pvnode.elm=nvn.elm},remove(){w4.destroy()},destroy(){w4.destroy()}}; w4.__widget__.pvnode = pvnode;});
    } else {
        def3 = def3.then(()=>{if (w4.__widget__.isDestroyed) {return};let vnode;if (!w4.__widget__.vnode){vnode=w4.__widget__.pvnode} else { vnode=h(w4.__widget__.vnode.sel, {key: 4});vnode.elm=w4.el;vnode.data.hook = {insert(a){a.elm.parentNode.replaceChild(w4.el,a.elm);a.elm=w4.el;w4.__mount();},remove(){w4.destroy()}, destroy() {w4.destroy()}}}c1[_2_index]=vnode;});
    }
    extra.promises.push(def3);
    return vn1;
}"
`;
